storage-kit/src/traits.rs

```diff
@@ -0,0 +1,128 @@
+use tokio::sync::mpsc;
+
+use crate::{FileChunk, FileKey, FileMetadata};
+
+/// Storage interface to be implemented by the storage providers.
+pub trait Storage: Clone + Send + Sync + 'static {
+    /// Get metadata for a file.
+    fn get_metadata(
+        &self,
+        key: &FileKey,
+    ) -> impl std::future::Future<Output = Option<FileMetadata>> + Send;
+
+    /// Set metadata for a file.
+    fn set_metadata(
+        &self,
+        key: &FileKey,
+        metadata: &FileMetadata,
+    ) -> impl std::future::Future<Output = ()> + Send;
+
+    /// Get a file chunk from storage.
+    fn get_chunk(
+        &self,
+        key: &FileKey,
+        chunk: u64,
+    ) -> impl std::future::Future<Output = Option<FileChunk>> + Send;
+
+    /// Write a file chunk in storage.
+    fn write_chunk(
+        &self,
+        key: &str,
+        chunk: u64,
+        data: &FileChunk,
+    ) -> impl std::future::Future<Output = ()> + Send;
+}
+
+/// The `Actor` trait represents an actor, which runs on its own event loop and can handle messages.
+/// The struct implementing this trait can be seen as the context of the actor, holding the internal
+/// state or the shared data (through commands and queries).
+pub trait Actor: Sized {
+    /// The type of message that the actor can handle.
+    /// Usually an enum that represents the different types of messages that the actor can receive.
+    type Message: Send + Sized + 'static;
+
+    /// The event loop associated with the actor.
+    /// If no custom event loop is needed, the default `EventLoop<Self>` can be used.
+    type EventLoop: ActorEventLoop<Self> + Send + 'static;
+
+    /// Handles a message received by the actor.
+    ///
+    /// * `message` - The message to be handled.
+    ///
+    fn handle_message(
+        &mut self,
+        message: Self::Message,
+    ) -> impl std::future::Future<Output = ()> + std::marker::Send;
+}
+
+/// Trait representing an event loop for an actor.
+pub trait ActorEventLoop<T: Actor> {
```

Comment:
This is just for composability. If you want to do something very specific (i.e. manage a network resource like `P2PModule`) you will have to implement your own event loop, otherwise you can use the provided one (`EventLoop`).